code/modules/clothing/accessories/holomap.dm

```diff
@@ -0,0 +1,255 @@
+
+	var/turf/last_turf = get_turf(src)
+	if(istype(loc, /obj/item/clothing/under))
+		var/obj/item/clothing/under/U = loc
+		if(U && ishuman(U.loc))
+			var/mob/living/carbon/human/H = U.loc
+			if(H.get_item_by_slot(slot_w_uniform) == U)
+				if(H && last_turf)
+					var/obj/item/clothing/accessory/holomap_chip/destroyed/D = new(last_turf)
+					D.marker_prefix = marker_prefix
+					D.holomap_filter = holomap_filter
+
+	deactivate_holomap()
+
+	for(var/cacheIcon in holomap_cache)
+		if(findtext(cacheIcon, "\ref[src]"))
+			holomap_cache -= cacheIcon
+	..()
+
+
+/obj/item/clothing/accessory/holomap_chip/can_attach_to(obj/item/clothing/C)
+	return (istype(C, /obj/item/clothing/under) && !C.action_button_name)
+
+
+/obj/item/clothing/accessory/holomap_chip/on_attached(obj/item/clothing/C)
+	..()
+	attached_to.action_button_name = "Toggle Holomap"
+
+
+/obj/item/clothing/accessory/holomap_chip/on_removed(mob/user as mob)
+	attached_to.action_button_name = null
+	deactivate_holomap()
+	..()
+
+
+/obj/item/clothing/accessory/holomap_chip/proc/togglemap()
+	if(usr.isUnconscious())
+		return
+
+	if(!attached_to)
+		return
+
+	if(!ishuman(usr))
+		to_chat(usr, "<span class='warning'>Only humanoids can use this device</span>")
+		return
+
+	var/mob/living/carbon/human/H = usr
+
+	if(!istype(loc))
+		to_chat(H, "<span class='warning'>This device needs to be set on a uniform first.</span>")
+
+	if(H.get_item_by_slot(slot_w_uniform) != attached_to)
+		to_chat(H, "<span class='warning'>You need to wear the suit first</span>")
+		return
+
+	if(activator)
+		deactivate_holomap()
+		to_chat(H, "<span class='notice'>You disable the holomap.</span>")
+	else
+		activator = H
+		processing_objects.Add(src)
+		process()
+		to_chat(H, "<span class='notice'>You enable the holomap.</span>")
+
+
+
+#define HOLOMAP_ERROR	0
+#define HOLOMAP_YOU		1
+#define HOLOMAP_OTHER	2
+#define HOLOMAP_DEAD	3
+
+/obj/item/clothing/accessory/holomap_chip/proc/update_holomap()
+	var/turf/T = get_turf(src)
+	if(!T)//nullspace begone!
+		return
+
+	if((!attached_to) || (!activator) || (activator.get_item_by_slot(slot_w_uniform) != attached_to) || (!activator.client) || (holoMiniMaps[T.z] == null))
+		deactivate_holomap()
+		return
+
+	activator.client.images -= holomap_images
+
+	holomap_images.len = 0
+
+	var/image/bgmap
+	var/holomap_bgmap
+
+	if(T.z == map.zCentcomm)
+		holomap_bgmap = "background_\ref[src]_[map.zCentcomm]"
+
+		if(!(holomap_bgmap in holomap_cache))
+			holomap_cache[holomap_bgmap] = image(centcommMiniMaps[holomap_filter])
+	else
+		holomap_bgmap = "background_\ref[src]_[T.z]"
+
+		if(!(holomap_bgmap in holomap_cache))
+			holomap_cache[holomap_bgmap] = image(holoMiniMaps[T.z])
+
+	bgmap = holomap_cache[holomap_bgmap]
+	bgmap.plane = HUD_PLANE
+	bgmap.layer = HUD_BASE_LAYER
+	bgmap.color = holomap_color
+	bgmap.loc = activator.hud_used.holomap_obj
+
+	if(!bgmap.pixel_x)
+		bgmap.pixel_x = -1*T.x + activator.client.view*WORLD_ICON_SIZE + 16*(WORLD_ICON_SIZE/32)
```

Comment:
I mean I get that, what I don't get is what is this default offset supposed to be or how its constructed, if you just put a few comments to tell people how its built it could save someone a huge headache in the future.